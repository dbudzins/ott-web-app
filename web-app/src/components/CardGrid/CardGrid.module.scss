.cell {
  padding: 4px;
}
.container {
  margin: 0 -4px;
}
