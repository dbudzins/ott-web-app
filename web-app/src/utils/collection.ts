import type { Consent, CustomerConsent } from '#types/account';
import type { Config } from '#types/Config';
import type { GenericFormValues } from '#types/form';
import type { Playlist, PlaylistItem } from '#types/playlist';

const getFiltersFromConfig = (config: Config, playlistId: string | undefined): string[] => {
  const menuItem = config.menu.find((item) => item.contentId === playlistId);
  const filters = menuItem?.filterTags?.split(',').filter(Boolean);

  return filters || [];
};

const filterPlaylist = (playlist: PlaylistItem[], filter: string) => {
  if (!filter) return playlist;

  return playlist.filter(({ tags }) => (tags ? tags.split(',').includes(filter) : false));
};

const chunk = <T>(input: T[], size: number) => {
  return input?.reduce((arr: T[][], item, idx: number) => {
    return idx % size === 0 ? [...arr, [item]] : [...arr.slice(0, -1), [...arr.slice(-1)[0], item]];
  }, []);
};

const findPlaylistImageForWidth = (playlistItem: PlaylistItem, width: number): string =>
  playlistItem.images.find((img) => img.width === width)?.src || playlistItem.image;

const generatePlaylistPlaceholder = (playlistLength: number = 15): Playlist => ({
  title: '',
  playlist: new Array(playlistLength).fill({}).map(
    (_value, index) =>
      ({
        description: '',
        duration: 0,
        feedid: '',
        image: '',
        images: [],
        link: '',
        genre: '',
        mediaid: `placeholder_${index}`,
        pubdate: 0,
        rating: '',
        sources: [],
        tags: '',
        title: '',
        tracks: [],
      } as PlaylistItem),
  ),
});

const formatConsentValues = (publisherConsents: Consent[] | null, customerConsents: CustomerConsent[] | null) => {
  if (!publisherConsents || !customerConsents) {
    return {};
  }

  const values: Record<string, boolean> = {};
  publisherConsents?.forEach((publisherConsent) => {
    if (customerConsents?.find((customerConsent) => customerConsent.name === publisherConsent.name && customerConsent.state === 'accepted')) {
      values[publisherConsent.name] = true;
    }
  });

  return values;
};

const extractConsentValues = (consents?: Consent[]) => {
  const values: Record<string, boolean> = {};

  if (!consents) {
    return values;
  }

  consents?.forEach((consent) => {
    values[consent.name] = consent.enabledByDefault;
  });

  return values;
};

const formatConsentsFromValues = (publisherConsents: Consent[] | null, values?: GenericFormValues) => {
  const consents: CustomerConsent[] = [];

  if (!publisherConsents || !values) return consents;

  publisherConsents.forEach((consent) => {
    consents.push({
      name: consent.name,
      version: consent.version,
      state: values.consents[consent.name] ? 'accepted' : 'declined',
    });
  });

  return consents;
};

const checkConsentsFromValues = (publisherConsents: Consent[], consents: Record<string, boolean>) => {
  const customerConsents: CustomerConsent[] = [];
  const consentsErrors: string[] = [];

  if (!publisherConsents || !consents) return { customerConsents, consentsErrors };

  publisherConsents.forEach((consent) => {
    if (consent.required && !consents[consent.name]) {
      consentsErrors.push(consent.name);
    }

    customerConsents.push({
      name: consent.name,
      version: consent.version,
      state: consents[consent.name] ? 'accepted' : 'declined',
    });
  });

  return { customerConsents, consentsErrors };
};

const deepCopy = (obj: unknown) => {
  if (Array.isArray(obj) || (typeof obj === 'object' && obj !== null)) {
    return JSON.parse(JSON.stringify(obj));
  }
  return obj;
};

export {
  getFiltersFromConfig,
  filterPlaylist,
  chunk,
  findPlaylistImageForWidth,
  generatePlaylistPlaceholder,
  formatConsentValues,
  formatConsentsFromValues,
  extractConsentValues,
  checkConsentsFromValues,
  deepCopy,
};
