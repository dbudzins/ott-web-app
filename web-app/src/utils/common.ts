export function debounce<T extends (...args: any[]) => void>(callback: T, wait = 200) {
  let timeout: NodeJS.Timeout | null;
  const callable = (...args: unknown[]) => {
    if (timeout) clearTimeout(timeout);
    timeout = setTimeout(() => callback(...args), wait);
  };
  return callable;
}

/**
 * Parse hex color and return the RGB colors
 * @param color
 * @return {{r: number, b: number, g: number}|undefined}
 */
export function hexToRgb(color: string) {
  if (color.indexOf('#') === 0) {
    color = color.slice(1);
  }

  // convert 3-digit hex to 6-digits.
  if (color.length === 3) {
    color = color[0] + color[0] + color[1] + color[1] + color[2] + color[2];
  }

  if (color.length !== 6) {
    return undefined;
  }

  return {
    r: parseInt(color.slice(0, 2), 16),
    g: parseInt(color.slice(2, 4), 16),
    b: parseInt(color.slice(4, 6), 16),
  };
}

/**
 * Get the contrast color based on the given color
 * @param {string} color Hex or RGBA color string
 * @link {https://stackoverflow.com/a/35970186/1790728}
 * @return {string}
 */
export function calculateContrastColor(color: string) {
  const rgb = hexToRgb(color);

  if (!rgb) {
    return '';
  }

  // http://stackoverflow.com/a/3943023/112731
  return rgb.r * 0.299 + rgb.g * 0.587 + rgb.b * 0.114 > 186 ? '#000000' : '#FFFFFF';
}

export const IS_DEV_BUILD = import.meta.env.DEV;
export const IS_TEST_BUILD = import.meta.env.MODE === 'test';

export function logDev(message: unknown, ...optionalParams: unknown[]) {
  if (IS_DEV_BUILD && !IS_TEST_BUILD) {
    console.info(message, optionalParams);
  }
}

export function getOverrideIP() {
  return document.cookie
    .split(';')
    .find((s) => s.trim().startsWith('overrideIP'))
    ?.split('=')[1]
    .trim();
}
