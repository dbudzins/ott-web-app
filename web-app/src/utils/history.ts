import type { History } from 'history';

export function addQueryParam(history: History, key: string, value: string): string {
  const urlSearchParams = new URLSearchParams(history.location.search);

  urlSearchParams.set(key, value);

  const searchParams = urlSearchParams.toString();

  return `${history.location.pathname}${searchParams ? `?${searchParams}` : ''}`;
}

export function removeQueryParam(history: History, key: string): string {
  const urlSearchParams = new URLSearchParams(history.location.search);

  urlSearchParams.delete(key);

  const searchParams = urlSearchParams.toString();

  return `${history.location.pathname}${searchParams ? `?${searchParams}` : ''}`;
}
