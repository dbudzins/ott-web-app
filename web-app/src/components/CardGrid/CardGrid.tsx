import React from 'react';
import type { GridCellProps } from 'react-virtualized';
import { isLocked } from 'ott-common/src/utils/entitlements';
import type { PlaylistItem } from 'ott-common/types/playlist';
import type { AccessModel } from 'ott-common/types/accessModel';

import styles from './CardGrid.module.scss';

import useBreakpoint, { Breakpoint, Breakpoints } from '#src/hooks/useBreakpoint';
import { chunk, findPlaylistImageForWidth } from '#src/utils/collection';
import Card from '#src/components/Card/Card';
import VirtualizedGrid from '#src/components/VirtualizedGrid/VirtualizedGrid';
import CardTag from '#src/components/Tag/CardTag';

const defaultCols: Breakpoints = {
  [Breakpoint.xs]: 2,
  [Breakpoint.sm]: 2,
  [Breakpoint.md]: 3,
  [Breakpoint.lg]: 4,
  [Breakpoint.xl]: 5,
};

type CardGridProps = {
  playlist: PlaylistItem[];
  onCardHover?: (item: PlaylistItem) => void;
  onCardClick: (item: PlaylistItem, playlistId?: string) => void;
  watchHistory?: { [key: string]: number };
  isLoading: boolean;
  enableCardTitles?: boolean;
  cols?: Breakpoints;
  currentCardItem?: PlaylistItem;
  currentCardLabel?: string;
  accessModel: AccessModel;
  isLoggedIn: boolean;
  hasSubscription: boolean;
};

function CardGrid({
  playlist,
  onCardClick,
  onCardHover,
  watchHistory,
  enableCardTitles = true,
  isLoading = false,
  cols = defaultCols,
  currentCardItem,
  currentCardLabel,
  accessModel,
  isLoggedIn,
  hasSubscription,
}: CardGridProps) {
  const breakpoint: Breakpoint = useBreakpoint();
  const isLargeScreen = breakpoint >= Breakpoint.md;
  const imageSourceWidth = 320 * (window.devicePixelRatio > 1 || isLargeScreen ? 2 : 1);
  const rows = chunk<PlaylistItem>(playlist, cols[breakpoint]);

  const cellRenderer = ({ columnIndex, rowIndex, style }: GridCellProps) => {
    if (!rows[rowIndex][columnIndex]) return;

    const playlistItem: PlaylistItem = rows[rowIndex][columnIndex];
    const { mediaid, title } = playlistItem;

    return (
      <div className={styles.cell} style={style} key={mediaid} role="row">
        <div role="cell">
          <Card
            key={mediaid}
            title={title}
            enableTitle={enableCardTitles}
            tag={<CardTag item={playlistItem} />}
            posterSource={findPlaylistImageForWidth(playlistItem, imageSourceWidth)}
            progress={watchHistory ? watchHistory[mediaid] : undefined}
            onClick={() => onCardClick(playlistItem, playlistItem.feedid)}
            onHover={typeof onCardHover === 'function' ? () => onCardHover(playlistItem) : undefined}
            loading={isLoading}
            isCurrent={currentCardItem && currentCardItem.mediaid === mediaid}
            currentLabel={currentCardLabel}
            isLocked={isLocked(accessModel, isLoggedIn, hasSubscription, playlistItem)}
          />
        </div>
      </div>
    );
  };

  return (
    <div className={styles.container}>
      <VirtualizedGrid rowCount={rows.length} cols={cols} cellRenderer={cellRenderer} spacing={enableCardTitles ? 50 : 4} />
    </div>
  );
}

export default CardGrid;
