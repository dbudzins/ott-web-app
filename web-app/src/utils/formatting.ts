import { getSeriesId, getSeriesIdFromEpisode, isEpisode, isLiveChannel, isSeriesPlaceholder } from '#src/utils/media';
import type { PlaylistItem } from '#types/playlist';
import type { Series } from '#types/series';

/**
 * @param duration Duration in seconds
 *
 * Calculates hours and minutes into a string
 * Hours are only shown if at least 1
 * Minutes get rounded
 *
 * @returns string, such as '2h 24m' or '31m'
 */

export const formatDuration = (duration: number): string | null => {
  if (!duration) return null;

  const hours = Math.floor(duration / 3600);
  const minutes = Math.round((duration - hours * 3600) / 60);

  const hoursString = hours ? `${hours}h ` : '';
  const minutesString = minutes ? `${minutes}m ` : '';

  return `${hoursString}${minutesString}`;
};

export const addQueryParams = (url: string, queryParams: { [key: string]: string | number | string[] | undefined | null }) => {
  const queryStringIndex = url.indexOf('?');
  const urlWithoutSearch = queryStringIndex > -1 ? url.slice(0, queryStringIndex) : url;
  const urlSearchParams = new URLSearchParams(queryStringIndex > -1 ? url.slice(queryStringIndex) : undefined);

  Object.keys(queryParams).forEach((key) => {
    const value = queryParams[key];

    // null or undefined
    if (value == null) return;

    if (typeof value === 'object' && !value?.length) return;

    const formattedValue = Array.isArray(value) ? value.join(',') : value;

    urlSearchParams.set(key, String(formattedValue));
  });
  const queryString = urlSearchParams.toString();

  return `${urlWithoutSearch}${queryString ? `?${queryString}` : ''}`;
};

export const slugify = (text: string, whitespaceChar: string = '-') =>
  text
    .toString()
    .toLowerCase()
    .replace(/\s+/g, '-')
    .replace(/[^\w-]+/g, '')
    .replace(/--+/g, '-')
    .replace(/^-+/, '')
    .replace(/-+$/, '')
    .replace(/-/g, whitespaceChar);

export const movieURL = (item: PlaylistItem, playlistId?: string | null, play = false) =>
  addQueryParams(`/m/${item.mediaid}/${slugify(item.title)}`, { r: playlistId, play: play ? '1' : null });

export const seriesURL = (item: PlaylistItem, playlistId?: string | null, play = false) => {
  const seriesId = getSeriesId(item);

  return addQueryParams(`/s/${seriesId}/${slugify(item.title)}`, {
    r: playlistId,
    play: play ? '1' : null,
  });
};

export const liveChannelsURL = (playlistId: string, channelId?: string, play = false) => {
  return addQueryParams(`/p/${playlistId}`, {
    channel: channelId,
    play: play ? '1' : null,
  });
};

export const episodeURL = (series: Series, episode?: PlaylistItem, play: boolean = false, playlistId?: string | null) =>
  addQueryParams(`/s/${series.series_id}/${slugify(series.title)}`, {
    e: episode?.mediaid,
    r: playlistId,
    play: play ? '1' : null,
  });

export const episodeURLFromEpisode = (item: PlaylistItem, seriesId: string, playlistId?: string | null, play: boolean = false) => {
  // generated URL does not match the canonical URL. We need the series playlist in order to generate the slug. For
  // now the item title is used instead. The canonical link isn't affected by this though.
  return addQueryParams(`/s/${seriesId}/${slugify(item.title)}`, {
    e: item.mediaid,
    r: playlistId,
    play: play ? '1' : null,
  });
};

export const cardUrl = (item: PlaylistItem, playlistId?: string | null, play: boolean = false) => {
  if (isLiveChannel(item)) {
    return liveChannelsURL(item.liveChannelsId, item.mediaid, play);
  }

  if (isEpisode(item)) {
    const seriesId = getSeriesIdFromEpisode(item);

    return seriesId ? episodeURLFromEpisode(item, seriesId, playlistId, play) : movieURL(item);
  }

  return isSeriesPlaceholder(item) ? seriesURL(item, playlistId, play) : movieURL(item, playlistId, play);
};

export const videoUrl = (item: PlaylistItem, playlistId?: string | null, play: boolean = false) =>
  addQueryParams(item.seriesId ? seriesURL(item, playlistId) : movieURL(item, playlistId), {
    play: play ? '1' : null,
  });

export const formatDate = (dateString: number) => {
  if (!dateString) return '';

  return new Date(dateString * 1000).toLocaleDateString();
};

export const formatPrice = (price: number, currency: string, country: string) => {
  return new Intl.NumberFormat(country, {
    style: 'currency',
    currency: currency,
  }).format(price);
};

export const formatVideoMetaString = (item: PlaylistItem, episodesLabel?: string) => {
  const metaData = [];

  if (item.pubdate) metaData.push(new Date(item.pubdate * 1000).getFullYear());
  if (!episodesLabel && item.duration) metaData.push(formatDuration(item.duration));
  if (episodesLabel) metaData.push(episodesLabel);
  if (item.genre) metaData.push(item.genre);
  if (item.rating) metaData.push(item.rating);

  return metaData.join(' â€¢ ');
};
