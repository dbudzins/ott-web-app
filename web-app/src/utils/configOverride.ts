import { IS_DEV_BUILD } from '#src/utils/common';

// In production, use local storage so the override persists indefinitely without the query string
// In dev mode, use session storage so the override persists until the tab is closed and then resets
const storage = IS_DEV_BUILD ? window.sessionStorage : window.localStorage;
const CONFIG_HOST = import.meta.env.APP_API_BASE_URL;
const INCLUDE_TEST_CONFIGS = import.meta.env.APP_INCLUDE_TEST_CONFIGS;

const configFileQueryKey = 'c';
const configFileStorageKey = 'config-file-override';

const DEFAULT_SOURCE = import.meta.env.APP_CONFIG_DEFAULT_SOURCE?.toLowerCase();
const ALLOWED_SOURCES = import.meta.env.APP_CONFIG_ALLOWED_SOURCES?.split(' ').map((source) => source.toLowerCase()) || [];
const UNSAFE_ALLOW_DYNAMIC_CONFIG = import.meta.env.APP_UNSAFE_ALLOW_DYNAMIC_CONFIG;

export function getConfig() {
  return formatSourceLocation(getConfigOverride() || DEFAULT_SOURCE);
}

export const setStoredConfig = (value: string) => {
  storage.setItem(configFileStorageKey, value);
};

export const getStoredConfig = () => {
  return storage.getItem(configFileStorageKey)?.toLowerCase();
};

export const clearStoredConfig = () => {
  storage.removeItem(configFileStorageKey);
};

function getConfigOverride() {
  const url = new URL(window.location.href);

  if (url.searchParams.has(configFileQueryKey)) {
    const configQuery = url.searchParams.get(configFileQueryKey)?.toLowerCase();

    // Strip the config file query param from the URL and history since it's stored locally,
    // and then the url stays clean and the user will be less likely to play with the param
    url.searchParams.delete(configFileQueryKey);
    window.history.replaceState(null, '', url.toString());

    // If the query param exists but the value is empty, clear the storage and allow fallback to the default config
    if (!configQuery) {
      clearStoredConfig();
      return undefined;
    }

    // If it's valid, store it and return it
    if (isValidConfigSource(configQuery)) {
      setStoredConfig(configQuery);
      return configQuery;
    }

    // Yes this falls through to look up the stored value if the query string is invalid and that's OK
  }

  const storedSource = getStoredConfig();

  // Make sure the stored value is still valid before returning it
  if (storedSource && isValidConfigSource(storedSource)) {
    return storedSource;
  }

  return undefined;
}

function isValidConfigSource(source: string) {
  // Dynamic values are valid as long as they are defined
  if (UNSAFE_ALLOW_DYNAMIC_CONFIG) {
    return !!source;
  }

  if (INCLUDE_TEST_CONFIGS && source.startsWith('test--')) {
    return true;
  }

  return ALLOWED_SOURCES.indexOf(source) >= 0;
}

function formatSourceLocation(source?: string) {
  if (!source) {
    return undefined;
  }

  if (source.match(/^[a-z,\d]{8}$/)) {
    return `${CONFIG_HOST}/apps/configs/${source}.json`;
  }

  if (INCLUDE_TEST_CONFIGS && source.startsWith('test--')) {
    return `/test-data/config.${source}.json`;
  }

  return source;
}

/***
 * If present, re-add the config override key to the query string of the share url,
 * so the url copied will pull up the same site config
 * @param href The URL to append to as a string (i.e. window.location.href)
 */
export function addConfigParamToUrl(href: string) {
  const config = getConfigOverride();
  const url = new URL(href);

  url.searchParams.delete(configFileQueryKey);

  if (config) {
    url.searchParams.append(configFileQueryKey, config);
  }

  return url.toString();
}
