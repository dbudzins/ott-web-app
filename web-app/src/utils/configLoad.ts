import merge from 'lodash.merge';
import type { AccessModel } from 'ott-common/types/accessModel';

import { calculateContrastColor } from '../utils/common';
import { getConfig } from '../utils/configOverride';
import loadConfig, { validateConfig } from '../services/config.service';
import { addScript } from '../utils/dom';
import { useConfigStore } from '../stores/ConfigStore';

import type { Config, Styling } from '#types/Config';

const defaultConfig: Config = {
  id: '',
  siteName: '',
  description: '',
  player: '',
  assets: {},
  content: [],
  menu: [],
  integrations: {
    cleeng: {
      id: null,
      useSandbox: true,
    },
  },
  styling: {
    footerText: '',
    shelfTitles: true,
  },
  features: {
    enableSharing: true,
  },
};

const setCssVariables = ({ backgroundColor, highlightColor, headerBackground }: Styling) => {
  const root = document.querySelector(':root') as HTMLElement;

  if (root && backgroundColor) {
    root.style.setProperty('--body-background-color', backgroundColor);
    root.style.setProperty('--background-contrast-color', calculateContrastColor(backgroundColor));
  }

  if (root && highlightColor) {
    root.style.setProperty('--highlight-color', highlightColor);
    root.style.setProperty('--highlight-contrast-color', calculateContrastColor(highlightColor));
  }
  if (root && headerBackground) {
    root.style.setProperty('--header-background', headerBackground);
    root.style.setProperty('--header-contrast-color', calculateContrastColor(headerBackground));
  }
};

const maybeInjectAnalyticsLibrary = (config: Config) => {
  if (!config.analyticsToken) return;

  return addScript('/jwpltx.js');
};

const calculateAccessModel = (config: Config): AccessModel => {
  const { id, monthlyOffer, yearlyOffer } = config?.integrations?.cleeng || {};

  if (!id) return 'AVOD';
  if (!monthlyOffer && !yearlyOffer) return 'AUTHVOD';
  return 'SVOD';
};

export const loadAndValidateConfig = async (
  onLoading: (isLoading: boolean) => void,
  onValidationError: (error: Error) => void,
  onValidationCompleted: (config: Config) => Promise<void>,
) => {
  onLoading(true);

  const configLocation = getConfig();

  if (!configLocation) {
    onValidationError(new Error('Config not defined'));
    return;
  }

  const config = await loadConfig(configLocation).catch((error) => {
    onValidationError(error);
  });

  if (!config) {
    onLoading(false);
    return;
  }

  await validateConfig(config)
    .then(async (configValidated) => {
      const configWithDefaults = merge({}, defaultConfig, configValidated);

      const accessModel = calculateAccessModel(configWithDefaults);

      useConfigStore.setState({
        config: configWithDefaults,
        accessModel,
      });

      setCssVariables(configValidated.styling);
      maybeInjectAnalyticsLibrary(config);
      await onValidationCompleted(config);
    })
    .catch((error: Error) => {
      onValidationError(error);
    });
};
