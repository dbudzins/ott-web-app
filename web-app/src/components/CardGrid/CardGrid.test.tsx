import * as React from 'react';
import { render } from '@testing-library/react';

import { generatePlaylistPlaceholder } from '../../utils/collection';

import CardGrid from './CardGrid';

describe('<CardGrid>', () => {
  it('renders and matches snapshot', () => {
    const placeholderData = generatePlaylistPlaceholder();
    const { container } = render(
      <CardGrid
        playlist={placeholderData.playlist}
        onCardHover={vi.fn()}
        onCardClick={vi.fn()}
        isLoading={false}
        accessModel={'SVOD'}
        isLoggedIn={true}
        hasSubscription={true}
      />,
    );

    expect(container).toMatchSnapshot();
  });
});
