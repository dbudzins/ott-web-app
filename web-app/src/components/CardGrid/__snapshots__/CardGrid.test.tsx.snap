// Vitest Snapshot v1

exports[`<CardGrid> > renders and matches snapshot 1`] = `
<div>
  <div
    class="container"
  >
    <div
      style="overflow: visible; width: 0px;"
    >
      <div
        aria-label="grid"
        aria-readonly="true"
        class="ReactVirtualized__Grid grid"
        role="grid"
        style="box-sizing: border-box; direction: ltr; height: auto; position: relative; width: 0px; will-change: transform; overflow-x: hidden; overflow-y: hidden;"
        tabindex="-1"
      />
    </div>
    <div
      class="resize-triggers"
    >
      <div
        class="expand-trigger"
      >
        <div
          style="width: 1px; height: 1px;"
        />
      </div>
      <div
        class="contract-trigger"
      />
    </div>
  </div>
</div>
`;
